File: recursion/recursion-examples/print_backwards_recursive.cs
/*
Note: This file is designed to be copied out and compiled on your machine. 
In order for it to compile properly you need to ensure that the project name is the same as the "namespace" in this file. 

To run this file you need to:
1. Copy the contents
2. Paste them into the C# IDE of your choice (Visual Studio, for example)
3. Compile the program
4. Run the program
*/

using System;

namespace isaac_code_samples
{
    class PrintBackwardsRecursive
    {
        static void Main(string[] args){
            PrintBackwards("I am a computer scientist");
        }

        // Prints a given string backwards
        public static void PrintBackwards(string phrase){
            if (phrase.Length == 1){
                Console.Write(phrase);
            } else {
                string new_phrase = phrase.Substring(1, phrase.Length - 1);
                PrintBackwards(new_phrase);
                Console.Write(phrase.Substring(0, 1));
            }
        }
    }
}

File: recursion/recursion-examples/python/print_backwards_recursive.py
# Raspberry Pi Foundation
# Developed to be used alongside Isaac Computer Science,
# part of the National Centre for Computing Education
# Usage licensed under CC BY-SA 4

def print_backwards(phrase):
    """Prints a given word or phrase backwards"""
    if len(phrase) == 1:
        print(phrase, end = "")
    else:
        new_phrase = phrase[1:]
        print_backwards(new_phrase)
        print(phrase[0], end = "")


# This code will run if this file is executed directly
# (i.e. not called by another program) 
if __name__ == '__main__':
    print_backwards("I am a computer scientist")
